app/src/main/java/com/github/steroidteam/todolist/database/TodoListRepository.java

```diff
@@ -221,37 +217,98 @@ public void updateTask(UUID todoListID, int index, Task updatedTask) {
                 .thenAccept(this.observedTodoList::postValue);
     }
 
-    private void setTagsList() {
-        List<UUID> tagsIds = observedTodoList.getValue().getTagsIds();
-        localDatabase.getTagsFromIds(tagsIds).thenAccept(tagsList -> listTags.setValue(tagsList));
+    public LiveData<List<Tag>> getLocalTags(UUID todoListID) {
+        setTagsLists(todoListID);
+        return localTags;
+    }
+
+    public LiveData<List<Tag>> getGlobalTags() {
+        localDatabase.getAllTags().thenAccept(tags -> globalTags.postValue(tags));
+        return globalTags;
     }
 
-    public List<Tag> getTags() {
-        setTagsList();
-        return listTags.getValue();
+    public LiveData<List<Tag>> getUnlinkedTags(UUID todoListID) {
+        setTagsLists(todoListID);
+        return unlinkedTags;
     }
 
-    public void putTag(UUID todoListID, Tag tag) {
+    public void putTagInTodolist(UUID todoListID, UUID tagId) {
         localDatabase
-                .putTag(tag)
-                .thenCompose(t -> localDatabase.putTagInList(todoListID, t.getId()))
+                .putTagInList(todoListID, tagId)
                 .thenCompose(str -> localDatabase.getTodoList(todoListID))
+                .thenApply(TodoList::sortByDate)
                 .thenAccept(observedTodoList::postValue)
+                .thenCompose(str -> localDatabase.getTagsFromList(todoListID))
+                .thenAccept(tags -> localTags.postValue(tags))
                 .thenAccept(
-                        str ->
-                                localDatabase
-                                        .getTagsFromList(todoListID)
-                                        .thenAccept(tags -> listTags.postValue(tags)));
+                        str -> {
+                            List<Tag> unlinked = globalTags.getValue();
+                            unlinked.removeAll(globalTags.getValue());
+                            unlinkedTags.postValue(unlinked);
+                        });
     }
 
-    public void destroyTag(UUID todoListID, Tag tag) {
+    public void putTag(Tag tag) {
+        localDatabase
+                .putTag(tag)
+                .thenCompose(str -> localDatabase.getAllTags())
+                .thenAccept(globalTags::postValue)
+                .thenAccept(
+                        str -> {
+                            List<Tag> unlinked = globalTags.getValue();
+                            unlinked.removeAll(localTags.getValue());
+                            unlinkedTags.postValue(unlinked);
+                        });
+
+        remoteDatabase.putTag(tag);
+    }
+
+    public void removeTagFromTodolist(UUID todoListID, UUID tagId) {
```

Comment:
I wonder if we are not starting to pile-up a lot of round-trips when dealing with the database.